File: 2023/19_Aplenty/workflows.py

# ======================================================================
# Aplenty
#   Advent of Code 2023 Day 19 -- Eric Wastl -- https://adventofcode.com
#
# Python implementation by Dr. Dean Earl Wright III
# ======================================================================

# ======================================================================
#                        w o r k f l o w s . p y
# ======================================================================
"Workflows for the Advent of Code 2023 Day 19 puzzle"

# ----------------------------------------------------------------------
#                                                                 import
# ----------------------------------------------------------------------
from collections import namedtuple

from workflow import Workflow

# ----------------------------------------------------------------------
#                                                              constants
# ----------------------------------------------------------------------
START = "in"
FINAL = "AR"
ACCEPT = "A"
REJECT = "R"
LESS = "<"
MORE = ">"

# ----------------------------------------------------------------------
#                                                                  Types
# ----------------------------------------------------------------------
Range = namedtuple('Range', 'low high')
State = namedtuple('State', 'name x m a s')


def len_range(rnge):
    "Return the length of the range"
    return 1 + rnge.high - rnge.low


def len_state(state):
    "Return the product of the ranges in the search state"
    return (len_range(state.x) * len_range(state.m) *
            len_range(state.a) * len_range(state.s))

# ======================================================================
#                                                              Workflows
# ======================================================================


class Workflows(object):   # pylint: disable=R0902, R0903, R0205
    "Object for Aplenty"

    def __init__(self, text=None, part2=False):

        # 1. Set the initial values
        self.part2 = part2
        if text is None:
            self.text = []
        else:
            self.text = text
        self.workflows = {}

        # 2. Process text (if any)
        if text is not None and len(text) > 0:
            self._process_text(text)

    def _process_text(self, text):
        "Assign values from text"

        # 0. Precondition axioms
        assert text is not None and len(text) > 0

        # 1. Loop for each line of the text
        for line in text:

            # 2. Create and add a workflow
            new_wf = Workflow(text=line, part2=self.part2)
            name = new_wf.name
            assert name not in self.workflows
            self.workflows[name] = new_wf

    def add_workflow(self, text):
        "Add an additional workflow"

        # 1. Add the text and new workflow
        self.text.append(text)
        new_wf = Workflow(text=text, part2=self.part2)
        name = new_wf.name
        assert name not in self.workflows
        self.workflows[name] = new_wf

        # 2. Return the name of the new workflow
        return name

    def evaluate(self, part):
        "Evaluate the part returning A or R"

        # 0. Precondition axioms
        assert START in self.workflows

        # 1. Start at the very beginning
        state = START
        states = [state]

        # 2. Loop until we reach a final state
        while state not in FINAL:

            # 3. Test the part against the specified workflow
            assert state in self.workflows, f"{state} not in workflows"
            state = self.workflows[state].evaluate(part)
            states.append(state)

        # 4. Add the terminal state (last of the states)
        return state

    def combinations(self):
        "Return the number of part combinations"

        # 1. Start at the beginning
        searches = [State(name=START,
                          x=Range(low=1, high=4000),
                          m=Range(low=1, high=4000),
                          a=Range(low=1, high=4000),
                          s=Range(low=1, high=4000))]
        result = 0

        # 2. Loop while there is something to search
        while searches:

            # 3. Get the most recent search
            search = searches.pop()
            ranges = {
                "x": search.x,
                "m": search.m,
                "a": search.a,
                "s": search.s,
            }

            # 4. Have we reached a terminal state?
            name = search.name
            if name == REJECT:
                continue
            if name == ACCEPT:
                result += len_state(search)
                continue

            # 5. Loop for the rules in the named workflow
            for rule in self.workflows[name].get_rules():
                rule_category = rule.category
                rule_test = rule.test
                rule_value = rule.value
                rule_branch = rule.branch

                # 6. Adjust the ranges based on the test (if any)
                new_ranges = ranges.copy()
                if rule_test == LESS:
                    new_ranges[rule_category] = \
                        Range(low=new_ranges[rule_category].low, high=rule_value - 1)
                    ranges[rule_category] = Range(low=rule_value, high=ranges[rule_category].high)
                elif rule_test == MORE:
                    new_ranges[rule_category] = \
                        Range(low=rule_value + 1, high=new_ranges[rule_category].high)
                    ranges[rule_category] = Range(low=ranges[rule_category].low, high=rule_value)

                # 7. Add a new search to explore
                searches.append(State(name=rule_branch,
                                      x=new_ranges["x"],
                                      m=new_ranges["m"],
                                      a=new_ranges["a"],
                                      s=new_ranges["s"]))

        # 8. Return the combinations of ratings
        return result


# ----------------------------------------------------------------------
#                                                  module initialization
# ----------------------------------------------------------------------


if __name__ == '__main__':
    pass

# ======================================================================
# end                     w o r k f l o w s . p y                    end
# ======================================================================

File: 2023/19_Aplenty/workflows.go
// ======================================================================
// Aplenty
//   Advent of Code 2023 Day 19 -- Eric Wastl -- https://adventofcode.com
//
// go implementation by Dr. Dean Earl Wright III
// ======================================================================

// ======================================================================
//                         w o r k f l o w s . g o
// ======================================================================
// A solver for the Advent of Code 2023 Day 19 puzzle

package main

// ----------------------------------------------------------------------
//                                                                imports
// ----------------------------------------------------------------------
import (
	"fmt"
	"strings"
)

// ----------------------------------------------------------------------
//
//	constants
//
// ----------------------------------------------------------------------
const (
	STARTAT  = "in"
	REJECTED = "R"
	ACCEPTED = "A"
)

// ======================================================================
//                                                              Workflows
// ======================================================================

// Workflows ... Object for Aplenty
type Workflows struct {
	Part2 bool
	Text  []string
	Flows map[string]*Workflow
	Parts []*Part
}

// NewWorkflows ... Create a Workflows object
func NewWorkflows(part2 bool, text []string) (*Workflows, error) {

	// 1. Set the initial values
	p := new(Workflows)
	p.Part2 = part2
	p.Text = text
	p.Flows = make(map[string]*Workflow)
	p.Parts = make([]*Part, 0)

	// 2. Process text (if any)
	if len(p.Text) > 0 {
		err := p.processText(p.Text)
		if err != nil {
			return nil, err
		}
	}

	// 3. Return the new Workflows object
	return p, nil
}

// processText ... Assign values from text
func (p *Workflows) processText(text []string) error {

	// 1. Loop for each line of the text
	for indx, line := range text {

		// 2. Lines that start with a left currly brace are parts
		if strings.HasPrefix(line, "{") {
			part, err := NewPart(p.Part2, line)
			if err != nil {
				return fmt.Errorf("error (part) in line %d: %s", indx, err)
			}
			p.Parts = append(p.Parts, part)
			continue
		}

		// 3. Opther lines are workflows
		workflow, err := NewWorkflow(p.Part2, line)
		if err != nil {
			return fmt.Errorf("error (workflow) in line %d: %s", indx, err)
		}
		p.Flows[workflow.Name] = workflow
	}

	// 4. Return success
	return nil
}

// CheckPart ... Check a part against the workflow, returning A or R
func (p *Workflows) CheckPart(part *Part) string {

	// 1. Start at the very beginning
	wfName := STARTAT

	// 2. Loop until resolution
	for {

		// 3. Process the current rule
		wfName = p.Flows[wfName].CheckPart(part)

		// 4. If we have reach the end, return result
		if wfName == REJECTED || wfName == ACCEPTED {
			return wfName
		}
	}
}

// CheckParts ... Check all the parts, total rating number of those approved
func (p *Workflows) CheckParts() int {

	// 1. Start with nothing
	result := 0

	// 2. Loop for all of the parts
	for _, part := range p.Parts {

		// 3. Evaluate the part
		evaluation := p.CheckPart(part)

		// 4. If approved, accumulating rating
		if evaluation == ACCEPTED {
			result = result + part.Rating()
		}
	}

	// 5. Return total rating number of approved parts
	return result
}

// Combinations
func (p *Workflows) Combinations(verbose bool, limit int) int {

	// 1. Start with nothing
	result := 0

	// 2. Initialize queue
	initialRange, _ := NewRanges(true, STARTAT)
	queue := []*Ranges{initialRange}

	// 3. Loop until the queue is empty
	for len(queue) > 0 {
		if verbose {
			fmt.Printf("Queue length %d, result %d\n", len(queue), result)
		}

		// 4. Get the next item to evaluate
		indx := len(queue) - 1
		item := queue[indx]
		queue = queue[:indx]
		if verbose {
			fmt.Printf("Evaluating %s\n", item.String())
		}

		// 5. Ignore if Rejected
		if item.Text == REJECTED {
			continue
		}

		// 6. Accumulate the combinations if Accepted
		if item.Text == ACCEPTED {
			result = result + item.Combinations()
			continue
		}

		// 7. Process the item
		more := p.Flows[item.Text].NextRanges(item)

		// 8. Add the new items to the queue
		queue = append(queue, more...)
	}

	// 9. Return the total number of combinations
	return result
}

// PartOne ... Returns the solution for part one
func (p *Workflows) PartOne(verbose bool, limit int) string {

	// 1. Return the solution for part one
	return fmt.Sprint(p.CheckParts())
}

// PartTwo ... Returns the solution for part two
func (p *Workflows) PartTwo(verbose bool, limit int) string {

	// 1. Return the solution for part two
	return fmt.Sprint(p.Combinations(false, 10000))
}

// ======================================================================
// end                     w o r k f l o w s . g o                    end
// ======================================================================
